### DISCUSSION TITLE
Using immer in a wxWidgets app

### DISCUSSION THREAD

**Discussion by richtw1** [arximboldi/immer#245] [Q&A]
I was wondering if anyone has tried using immer in a wxWidgets app, and whether they were able to implement the unidirectional dataflow architecture (as per the "canonical" example from Juanpe's 2017 talk):

```
auto term = terminal{};
auto state = application{};
while (auto new_state = update(state, term.next())) {
    state = *new_state;
    draw(state);
}
```

Obviously this is at odds with the wxWidgets architecture, which implements the event loop as an internal detail, and exposes customization points for each possible event.

Currently my wxApplication class holds the application state as a member variable, and my derived widget classes hold references to the wxApplication class so their event handlers can mutate that state.  This doesn't feel much in the spirit of value oriented design, but more like a lazy compromise in order to work around the expectations of the wxWidgets API.

Has anyone found a neater solution?  wxWidgets will let you write a custom event loop, but this would also require a whole layer to interface between the wxWidgets event/widget API and a new immer-centric API where events return a value which can be passed to `update()`.

Probably some people will suggest that I use dear imgui or similar, but I'm really after a "native look and feel" and wxWidgets is the only framework I've found which can do that in a cross-platform way.

**Comment by arximboldi** [comment]
That example is just to illustrate the conceptual logic, but Lager is designed to interact with a more modern "event loop" system. 

The `make_store()` function takes an event loop interface as a parameter. This interfaces allows Lager to interact with the UI system's event loop: when you dispatch an action, it schedules a reduction of the action into the event loop. Effects execution is also achieved by scheduling them into the event loop.

Lager itself provides a few of these "event loop" implementations. You can check `lager/event_loop/...` folder for examples that work with Qt, SDL, and others. I do not know exactly how the wxWidgets API is, but I can imagine writing such an event loop interface for it should be rather easy. The main thing that you need is some function to "schedule" an arbitrary function into the event loop.

You can try inspire yourself using the existing code as basis and try to do it. I can answer further questions if you have some. I'd also be happy to accept a PR that adds wxWidgets support if you wanna!

---

PS. I couldn't resist myself and took a quick look at how the approach would be. It seems to me that a viable approach that reuses the existings mechanism would be to use the existing `safe_queue_event_loop`, and run the event queue (`step()`) in the `MyApp::OnIdle()` event of the wxWidgets application class. The `safe_queue_event_loop` allows you to use `dispatch()` from any thread -- if your application is single threaded you can also just go with `queue_event_loop`, but generally the multi-threaded `dispatch()` tends to be super useful.
